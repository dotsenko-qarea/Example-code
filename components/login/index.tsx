import React from 'react';
import { Platform, Pressable, Text, View } from 'react-native';

import { Controller } from 'react-hook-form';
import DeviceInfo from 'react-native-device-info';
import { KeyboardAwareScrollView } from 'react-native-keyboard-aware-scroll-view';
import { Button, TextInput } from 'react-native-paper';

import tw from '~modules/lib/tailwind';

import { HeaderLogin } from '../header';
import { useLogin } from '../hooks';

export const Login: React.FC = () => {
  const { handleShowPassword, handleAdminInfo, onSubmit, errors, control, handleSubmit, isShowPassword, isOnline } =
    useLogin();

  return (
    <View style={tw`justify-between h-full`}>
      <KeyboardAwareScrollView
        enableOnAndroid
        extraScrollHeight={Platform.OS === 'ios' ? 25 : 60}
        enableAutomaticScroll
        keyboardShouldPersistTaps='handled'
      >
        <HeaderLogin online={isOnline || false} />
        <View style={tw`px-4 mt-14`}>
          <Text style={tw`text-primary mb-8 text-xl`}>{`Some Text`}</Text>
          <View style={tw`mb-4 h-18`}>
            <Controller
              control={control}
              rules={{
                required: true,
              }}
              render={({ field: { onChange, onBlur, value } }) => (
                <TextInput
                  keyboardType='email-address'
                  label='Email'
                  mode='outlined'
                  onBlur={onBlur}
                  onChangeText={(val: string) => onChange(val.toLocaleLowerCase().trim())}
                  value={value}
                  error={!!errors?.email}
                />
              )}
              name='email'
            />
            {errors.email && <Text style={tw`text-error`}>{errors.email.message}</Text>}
          </View>
          <View style={tw`mb-4 h-18`}>
            <Controller
              control={control}
              rules={{
                maxLength: 100,
              }}
              render={({ field: { onChange, onBlur, value } }) => (
                <TextInput
                  secureTextEntry={!isShowPassword}
                  label='Password'
                  mode='outlined'
                  textContentType='password'
                  onBlur={onBlur}
                  onChangeText={onChange}
                  value={value}
                  error={!!errors?.password}
                  right={
                    <TextInput.Icon
                      name={isShowPassword ? 'eye-off-outline' : 'eye-outline'}
                      onPress={handleShowPassword}
                    />
                  }
                />
              )}
              name='password'
            />
            {errors.password && <Text style={tw`text-error`}>{errors.password.message}</Text>}
          </View>
          <Button
            mode='contained'
            style={tw`mt-2 rounded`}
            labelStyle={tw`text-base font-semibold`}
            onPress={handleSubmit(onSubmit)}
          >
            {`Login`}
          </Button>
          <Pressable onPress={handleAdminInfo} style={tw`mt-10`}>
            <Text style={tw`text-lg text-primary`}>{`Some Text`}</Text>
          </Pressable>
        </View>
      </KeyboardAwareScrollView>
      <Text style={tw`text-center`}>{`Version: ${DeviceInfo.getVersion()}`}</Text>
    </View>
  );
};
