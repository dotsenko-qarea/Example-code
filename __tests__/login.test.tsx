import React from 'react';

import mockAsyncStorage from '@react-native-async-storage/async-storage/jest/async-storage-mock';
import renderer from 'react-test-renderer';
import { RecoilRoot } from 'recoil';

import { HeaderLogin, Login, LoginScreen } from '~modules/login';
import { RealmProvider } from '~modules/realm';

jest.mock('react-native-device-info', () => {
  return {
    getVersion: () => 4,
  };
});
jest.mock('@react-navigation/core');
jest.mock('react-native-keyboard-aware-scroll-view');
jest.mock('react-native-safe-area-context', () => {
  const inset = { top: 0, right: 0, bottom: 0, left: 0 };

  return {
    SafeAreaProvider: jest.fn().mockImplementation(({ children }) => children),
    SafeAreaConsumer: jest.fn().mockImplementation(({ children }) => children(inset)),
    useSafeAreaInsets: jest.fn().mockImplementation(() => inset),
  };
});
jest.mock('@react-navigation/native', () => ({
  ...jest.requireActual('@react-navigation/native'),
  useNavigation: () => ({ goBack: jest.fn() }),
}));
jest.mock('react-native/Libraries/Animated/NativeAnimatedHelper');
jest.mock('@react-native-async-storage/async-storage', () => mockAsyncStorage);

describe('modules', () => {
  describe('login', () => {
    it('renders <LoginScreen />', () => {
      const loginScreen = renderer
        .create(
          <RecoilRoot>
            <RealmProvider>
              <LoginScreen />
            </RealmProvider>
          </RecoilRoot>
        )
        .toJSON();

      expect(loginScreen).toMatchSnapshot();
    });

    it('renders <HeaderLogin />', () => {
      const headerLogin = renderer.create(<HeaderLogin online={false} />).toJSON();

      expect(headerLogin).toMatchSnapshot();
    });

    it('renders <Login />', () => {
      const login = renderer
        .create(
          <RecoilRoot>
            <RealmProvider>
              <Login />
            </RealmProvider>
          </RecoilRoot>
        )
        .toJSON();

      expect(login).toMatchSnapshot();
    });
  });
});
