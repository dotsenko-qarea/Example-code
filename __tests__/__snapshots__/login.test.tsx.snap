// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`modules login renders <HeaderLogin /> 1`] = `
[
  <View
    style={
      {
        "alignItems": "center",
        "backgroundColor": "#2d9bf0",
        "flexDirection": "column",
        "paddingBottom": 16,
        "paddingLeft": 16,
        "paddingRight": 16,
        "paddingTop": 48,
      }
    }
  >
    <View
      style={
        {
          "alignItems": "center",
          "flexDirection": "row",
          "justifyContent": "center",
          "marginBottom": 8,
          "width": "100%",
        }
      }
    >
      <Text
        allowFontScaling={false}
        selectable={false}
        style={
          [
            {
              "color": "#f14725",
              "fontSize": 12,
            },
            undefined,
            {
              "fontFamily": "Ionicons",
              "fontStyle": "normal",
              "fontWeight": "normal",
            },
            {},
          ]
        }
      >
        
      </Text>
      <Text
        style={
          {
            "color": "#f14725",
            "fontSize": 12,
            "lineHeight": 16,
            "marginLeft": 8,
          }
        }
      >
        Відсутнє з'єднання з інтернетом
      </Text>
    </View>
    <View
      style={
        {
          "alignItems": "center",
          "flexDirection": "row",
          "justifyContent": "space-between",
          "width": "100%",
        }
      }
    >
      <Image
        accessibilityLabel="logo"
        source={
          {
            "testUri": "../../../assets/images/sources/logo.png",
          }
        }
        style={
          [
            {
              "width": 50,
            },
            {
              "height": 50,
            },
            undefined,
          ]
        }
      />
      <Text
        style={
          {
            "color": "#fff",
            "fontSize": 18,
            "lineHeight": 28,
            "textAlign": "center",
          }
        }
      >
        Державна прикордонна 
 Служба України
      </Text>
      <View
        style={
          {
            "width": 40,
          }
        }
      />
    </View>
  </View>,
  <View
    style={
      {
        "alignItems": "center",
        "backgroundColor": "#fff544",
        "flexDirection": "row",
        "justifyContent": "space-between",
        "paddingBottom": 16,
        "paddingLeft": 16,
        "paddingRight": 16,
        "paddingTop": 16,
      }
    }
  >
    <Image
      accessibilityLabel="coatOfArms"
      source={
        {
          "testUri": "../../../assets/images/sources/coatOfArms.png",
        }
      }
      style={
        [
          {
            "width": 50,
          },
          {
            "height": 70,
          },
          undefined,
        ]
      }
    />
    <Text
      style={
        {
          "color": "#2268a4",
          "fontSize": 30,
          "fontWeight": "bold",
          "lineHeight": 36,
          "textAlign": "center",
        }
      }
    >
      Border register
    </Text>
    <View
      style={
        {
          "width": 40,
        }
      }
    />
  </View>,
]
`;

exports[`modules login renders <Login /> 1`] = `null`;

exports[`modules login renders <LoginScreen /> 1`] = `null`;
